packages/headless/src/app/ssr-engine/ssr-engine.ts

```diff
@@ -0,0 +1,465 @@
+/**
```

Comment:
> Didn't find the need for `search-engine` types file

My concern with this one, is that we're only supporting the SearchEngine for now, but we'll eventually need to support the ProductListingEngine, and maybe more.

> The dep graph is not as clean as you have outlined - e.g. `hydrate-initial-state` types file has `import` `{EngineDefinition} from './core-engine-ssr-types';` because `EngineDefinition` depends on multiple types that cannot be moved into common. This is the only exception for now from what I see.

Is it only because of `InferHydrationResult`? If so, I think we should move that particular type to `core-engine-ssr-types`, since it's a circular dependency anyway.

Alternatively, you could replace `InferHydrationResult` to instead be:
```ts
type InferHydrationResult<
  T extends HydrateInitialStateWithProps<
    CoreEngine,
    ControllersMap,
    AnyAction,
    ControllersPropsMap
  >
> = Awaited<ReturnType<T['hydrateInitialState']>>;
```

> Since the files are nested under `ssr-engine/types/` wonder if the ..`ssr-types.ts` suffix is redundant?

Good point. I would be fine removing those suffix. Even `ssr` seems redundant in `ssr-engine.ts` honestly, since it's in the `ssr-engine` directory.

Reply:
* Create `search-engine` file 
* Removed redundant suffix from file names
* Reg moving `EngineDefinition` from core-engine to common, the issue is `EngineDefinitionWithProps`, `EngineDefinitionWithoutProps` which depend on `FetchInitialStateWithoutProps`, `HydrateInitialStateWithoutProps` and `BuildWithoutProps` ... so moving all of them into common would make common very long and other files too sparse.

Reply:
> * Reg moving `EngineDefinition` from core-engine to common, the issue is `EngineDefinitionWithProps`, `EngineDefinitionWithoutProps` which depend on `FetchInitialStateWithoutProps`, `HydrateInitialStateWithoutProps` and `BuildWithoutProps` ... so moving all of them into common would make common very long and other files too sparse.

I'm not sure what you mean by this.

I think `EngineDefinition` should stay in `core-engine`, but `InferHydrationResult` should either be moved to `core-engine` or be modified as to not need `EngineDefinition` (like my example).

Reply:
`EngineDefinition` depends on `EngineDefinitionWithProps` and `EngineDefinitionWithoutProps` which in turn depend on FetchInitialStateWithoutProps, HydrateInitialStateWithoutProps and BuildWithoutProps